### PR TITLE
refactor(bazel): Update Docker hashes as artifacts

### PR DESCRIPTION
**Pull request by zharinov**
## Changes

Move hash updating to a separate processing phase, i.e. artifact update.

## Context

- Ref: #9667

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by zharinov** [comment]
Example: https://github.com/renovate-testing/bazel-7/pull/1

File: lib/modules/manager/types.ts [hunk 1]
```diff
@@ -200,7 +202,7 @@ export interface UpdateArtifactsResult {
 
 export interface UpdateArtifact<T = Record<string, unknown>> {
   packageFileName: string;
-  updatedDeps: PackageDependency<T>[];
+  updatedDeps: Upgrade<T>[];
```
Comment (zharinov) [inline lib/modules/manager/types.ts:205]: Without this change, I'm unable to access `newDigest` field 🤷

Comment (rarkins) [reply lib/modules/manager/types.ts:205]: This interface is already used by a lot of other manager implementations of artifacts updating so we should think about the implications.

File: lib/modules/manager/bazel/update.ts [hunk 2]
```diff
@@ -121,10 +121,6 @@ export async function updateDependency({
         result = updateCode(result, [idx, 'tag'], newValue);
       }
 
-      if (newDigest) {
-        result = updateCode(result, [idx, 'digest'], newDigest);
-      }
```
Comment (rarkins) [inline lib/modules/manager/bazel/update.ts:126]: Can this digest be changed independently of the value? or we assume it's immutably tied to the value/release?

Comment (zharinov) [reply lib/modules/manager/bazel/update.ts:126]: Well, according [documentation](https://github.com/bazelbuild/rules_docker/blob/master/docs/container.md#container_pull), both `tag` and `digest` are optional, so I think it's better to treat it as independently updatable.

But can we infer this knowledge from the `update` structure? For example, when `newValue !== currentValue`, or using some other criteria.

Comment (rarkins) [reply lib/modules/manager/bazel/update.ts:126]: If tag and digest are independently updateable, they should be updated as part of the autoreplace (Eventually). i.e. part of update, not artifacts

Artifacts should update things which are dependent on the value/digest

Comment (zharinov) [reply lib/modules/manager/bazel/update.ts:126]: How can I tell apart one case from another reliably?

From the current `update.ts` code, they're independently updatable, but I don't trust it. If I remove `tag` parameter from `WORKSPACE`, the `hash` value should be updated in `update.ts`. But when both `tag` and `hash` are present, I think `hash` depends on `tag` value.

I'm having problem with the the criteria itself.

Comment (zharinov) [reply lib/modules/manager/bazel/update.ts:126]: To me, the course of action seems like we need to see how Renovate creates update with and without `tag` value, so we can spot the difference and see which condition to use for deciding whether it's artifact upgrade or the regular one.

Comment (rarkins) [reply lib/modules/manager/bazel/update.ts:126]: Can the digest update without the tag changing? If so then it's part of update not artifacts (and should work with autoreplace)

File: lib/modules/manager/types.ts (hunk 1 above)
Comment (zharinov) [reply lib/modules/manager/types.ts:205]: All I wanted is the field that is already present in this object, though `PackageDependency` interface prevents its usage. We receive `Upgrade<T>` in `updateDependency`, I see no reason why we can't have it in `updateArtifacts` as well.

Comment (rarkins) [reply lib/modules/manager/types.ts:205]: I think this Pr might need closing anyway? Digest needs to be updated the non artifacts way except when tags are immutable

Comment (zharinov) [reply lib/modules/manager/types.ts:205]: Tags are not guaranteed to be immutable, at least for GitHub, Docker and Go, isn't it?

**Comment by rarkins** [comment]
I renamed #9667 to reflect the true purpose. Let's close this and start over

**Pull request closed without merging**